### PR TITLE
Add Option to use Terraform Configuration in TestStep

### PR DESCRIPTION
**Pull request by bendbennett**
Closes: #150 

This PR adds the following fields to `TestStep`:

- `ConfigDirectory`
- `ConfigFile`
- `ConfigVariable`

### ConfigDirectory
Allows the specification of a path to a directory containing Terraform configuration files. All files in the directory are copied to the working directory used during the execution of an acceptance test `TestStep`.

### ConfigFile
Allows the specification of a path to a file containing Terraform configuration. The file is copied to the working directory used during the execution of an acceptance test `TestStep`.

### ConfigVariables
Allows defining Terraform variables for use in conjunction with Terraform configuration used during the execution of an acceptance test `TestStep`.

### CODE REVIEW

File: internal/teststep/config.go
Comment (bendbennett) [inline internal/teststep/config.go:26]: `HasDirectory()`, `HasRaw()`, `Raw()` and `WriteDirectory()` can all be removed and replaced with `Write(dest string)` once the logic for writing the raw configuration derived from `TestStep.Config` has been moved out of `*WorkingDir.SetConfig()`.

File: internal/teststep/config.go
Comment (bflad) [inline internal/teststep/config.go:188]: [`internal/plugintest`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-testing/internal/plugintest) has some helpers for this sorta thing -- if it makes more sense, maybe that functionality can split into its own internal file handling package to help prevent duplicating logic, but totally optional of course.

File: internal/teststep/config.go
Comment (bflad) [inline internal/teststep/config.go:97]: 💡 One thing you could do here to avoid checking each case repetitively introduce two separate `configuration` implementations -- `ConfigurationDirectory` and `ConfigurationString`

File: config/directory.go
Comment (bflad) [inline config/directory.go:20]: I think we may be able to add `TestName string` as well since the "front door" `resource.Test()` and friends should already be accepting a `testing.T`. 👍 I'll put a reciprocal comment on the function as well.

File: config/directory.go
Comment (bflad) [inline config/directory.go:48]: If we can thread `t.Name()` through to `TestStepConfigRequest` like I think we can, we should be able to drop the `testing.T` here. It'd also be awesome to show an example of what happens from the developer perspective in the Go documentation. 👍 

One 🌶️  take that I have is that I think we should encourage folks to use the Go conventional `testdata` directory as well, which I'm showing in the suggestion here. Reference: https://pkg.go.dev/cmd/go#hdr-Test_packages / https://github.com/golang/go/issues/14715 / https://go-review.googlesource.com/c/go/+/23353/3/src/cmd/go/test.go

```suggestion
// TestNameDirectory returns the name of the test when TestStepConfigFunc
// is executed. This facilitates a convention of naming directories
// containing Terraform configuration files with the name of the test.
//
// For example, given test code:
//
//    func TestExampleCloudThing_basic(t *testing.T) {
//        resource.Test(t, resource.TestCase{
//            Steps: []resource.TestStep{
//                {
//                    ConfigDirectory: config.TestNameDirectory(),
//                },
//            },
//        })
//    }
//
// The testing configurations will be expected in the
// testdata/TestExampleCloudThing_basic/ directory.
func TestNameDirectory() func(TestStepConfigRequest) string {
	return func(req TestStepConfigRequest) string {
		return filepath.Join("testdata", req.TestName)
	}
}
```

**Review by bflad — COMMENTED** [review]
Nice work so far, @bendbennett! I left some initial thoughts here. I really like where you are going with the new configuration handling interface and methods since that will help clean up the testing logic implementation.

File: config/variable.go
Comment (bendbennett) [reply config/variable.go]: Good call. Now using `json.Marshal(v)` directly.

File: config/variable.go
Comment (bendbennett) [reply config/variable.go]: Have added a comment. Happy to expand/reduce etc.

File: config/variable.go
Comment (bendbennett) [reply config/variable.go]: Have implemented `FloatVariable` and `IntegerVariable`. Have punted on `NumberVariable` for now but happy to add.

File: config_directory_test/config_directory_test.go
Comment (bendbennett) [reply config_directory_test/config_directory_test.go:1]: I believe this is now resolved via the following code change for detection of symlinks:

```go
if dirEntry.IsDir() || fi.Mode()&fs.ModeSymlink == fs.ModeSymlink {
```

File: internal/teststep/config.go
Comment (bendbennett) [reply internal/teststep/config.go]: Done 👍

File: internal/teststep/config.go
Comment (bendbennett) [reply internal/teststep/config.go]: Have added a `ConfigurationRequest.Validate()` method as per your suggestion.

File: internal/teststep/config.go
Comment (bendbennett) [reply internal/teststep/config.go:85]: I've left this as is for now if that's ok? Once these changes are merged I'll add an issue to consider refactoring.

File: internal/teststep/config.go
Comment (bendbennett) [reply internal/teststep/config.go]: I've added `configurationDirectory` and `configurationString` types that implement the `Config` interface. The `Configuration()` constructor function now returns `Config` interface to accommodate the various types we will need (e.g., `configurationDirectory`, `configurationString`, `configurationFile`). Owing to the switch to returning an interface from the `Configuration()` constructor function, nil checks have been added.

File: config/directory.go
Comment (bendbennett) [reply config/directory.go:19]: Added.

File: internal/teststep/directory_test.go
Comment (austinvalle) [inline internal/teststep/directory_test.go:60]: TIL this is valid 😆

File: website/docs/plugin/testing/acceptance-tests/configuration.mdx
Comment (austinvalle) [inline website/docs/plugin/testing/acceptance-tests/configuration.mdx:1]: Do we want to callout this documentation page on the existing `/terraform/plugin/testing/acceptance-tests/teststep#test-modes` page when talking about `Lifecycle` mode? 

All the examples in that file use the raw `Config`, but that's probably fine

File: website/docs/plugin/testing/acceptance-tests/configuration.mdx
Comment (austinvalle) [inline website/docs/plugin/testing/acceptance-tests/configuration.mdx:56]: Is the intention for developers to use the `required_providers` instead? Might be worth mentioning this doc: https://developer.hashicorp.com/terraform/language/providers/requirements#requiring-providers

File: website/docs/plugin/testing/acceptance-tests/configuration.mdx
Comment (austinvalle) [inline website/docs/plugin/testing/acceptance-tests/configuration.mdx:139]: Same note as above: https://developer.hashicorp.com/terraform/language/providers/requirements#requiring-providers

**Review by austinvalle — APPROVED** [review]
So much great work Ben! 🚀  - Couple notes and one thought here:

- With the amount of file/path specific tests we have now in `plugin-testing`, should we update our CI to run on other platforms as well (mac/windows)?

File: helper/resource/teststep_providers_test.go
Comment (bendbennett) [reply helper/resource/teststep_providers_test.go:2329]: Have added a test to demonstrate that the TestStep number is iterating and using config from the appropriately numbered directory.

File: internal/plugintest/util.go
Comment (bendbennett) [reply internal/plugintest/util.go:139]: Good spot! The `else if` is redundant and actually wouldn't work as shown. Have removed.

File: internal/teststep/directory_test.go
Comment (bendbennett) [reply internal/teststep/directory_test.go:60]: Me too 😆

File: website/docs/plugin/testing/acceptance-tests/configuration.mdx
Comment (bendbennett) [reply website/docs/plugin/testing/acceptance-tests/configuration.mdx:56]: Good point. Have added a sentence to make this clearer and used the link you supplied. Thanks

File: website/docs/plugin/testing/acceptance-tests/configuration.mdx
Comment (bendbennett) [reply website/docs/plugin/testing/acceptance-tests/configuration.mdx:139]: Good point. Have added a sentence to make this clearer and used the link you supplied. Thanks

**Comment by bendbennett** [comment]
> * With the amount of file/path specific tests we have now in `plugin-testing`, should we update our CI to run on other platforms as well (mac/windows)?

We could certainly consider this. Couple of observations when playing around with this:

- Updating the OS matrix in `ci-go.yml` to use windows-latest and macos-latest results in the error below when running on Windows. This can be handled by changing the running of the tests to something like `go test ./... -v -cover`:

```
no required module provides package .out; to add it: .........
```

- Next up, there's an [error](https://github.com/hashicorp/terraform-plugin-testing/actions/runs/5712009551/job/15474925149) generated for the `TestTest_TestStep_ProviderFactories_CopyWorkingDir_EachTestStep` test that's specific to MacOS 12. This can be handled by using `macos-13` in the testing matrix, but will require that we circle back round and update to `macos-latest` once the default for MacOS for GitHub actions has been bumped to 13.

- Finally, the tests that purposely fail to find a directory/file, for instance, `TestConfigurationFile_HasProviderBlock` "not-file", will need to be updated as the error message on Windows differs from Ubuntu and MacOS (e.g., `The system cannot find the file specified` versus `no such file or directory`).

I'm happy to make these changes if everyone is on board with the above or some variant thereof.

**Review by bflad — COMMENTED** [review]
Drive-by question (sorry I'm having trouble getting time to review this properly): Is there something handling the removal of configuration files not present when there are multiple `ConfigDirectory` steps?

e.g. if first test step has `ConfigDirectory` with `a.tf` and `b.tf.json` and second test step has `ConfigDirectory` has `c.tf` and `d.tf.json`, the working directory should only have `c.tf` and `d.tf.json` during the second test step.

File: helper/resource/teststep_providers_test.go
Comment (austinvalle) [inline helper/resource/teststep_providers_test.go:2338]: 🤣 🤣

**Comment by austinvalle** [comment]
> > * With the amount of file/path specific tests we have now in `plugin-testing`, should we update our CI to run on other platforms as well (mac/windows)?
> 
> We could certainly consider this. Couple of observations when playing around with this:
> 
> * Updating the OS matrix in `ci-go.yml` to use windows-latest and macos-latest results in the error below when running `go test -coverprofile=coverage.out ./...` on Windows. This can be handled by changing the running of the tests to something like `go test ./... -v -cover`:
> 
> ```
> no required module provides package .out; to add it: .........
> ```
> 
> * Next up, there's an [error](https://github.com/hashicorp/terraform-plugin-testing/actions/runs/5712009551/job/15474925149) generated for the `TestTest_TestStep_ProviderFactories_CopyWorkingDir_EachTestStep` test that's specific to MacOS 12. This can be handled by using `macos-13` in the testing matrix, but will require that we circle back round and update to `macos-latest` once the default MacOS for GitHub actions has been bumped to 13.
> * Finally, the tests that purposely fail to find a directory/file, for instance, `TestConfigurationFile_HasProviderBlock` "not-file", will need to be updated as the error message on Windows differs from Ubuntu and MacOS (e.g., `The system cannot find the file specified` versus `no such file or directory`).
> 
> I'm happy to make these changes if everyone is on board with the above or some variant thereof.

Hmm 🤔 , I don't want it to be too big of a pain... Maybe we just skip running them for different OS's for now? If we start running into bug reports that are OS specific we can revisit. I'd imagine the amount of users running into Window specific bugs is limited for Go modules like this anyways 🤷🏻

**Comment by bendbennett** [comment]
> Drive-by question (sorry I'm having trouble getting time to review this properly): Is there something handling the removal of configuration files not present when there are multiple `ConfigDirectory` steps?
> 
> e.g. if first test step has `ConfigDirectory` with `a.tf` and `b.tf.json` and second test step has `ConfigDirectory` has `c.tf` and `d.tf.json`, the working directory should only have `c.tf` and `d.tf.json` during the second test step.

Good point! Have added something to handle removal of config and variables files from previous test step prior to running the current test step along with some test coverage.

File: config/directory.go
Comment (bendbennett) [reply config/directory.go]: Have updated the docs to this effect.

File: config/directory.go
Comment (bendbennett) [reply config/directory.go]: Have updated the docs to this effect.

File: internal/plugintest/working_dir.go
Comment (bflad) [inline internal/plugintest/working_dir.go:103]: Open question: Not sure if we should or shouldn't be concerned about non-Terraform JSON files here.

**Review by bflad — APPROVED** [review]
Overall looks good to me 🚀

File: internal/plugintest/working_dir.go
Comment (bendbennett) [reply internal/plugintest/working_dir.go:103]: I _think_ that the only `.json` files within the `working_dir` will be those that are copied when a `.json` file, or a directory containing `.json` files is specified for `TestStep.ConfigFile`, or`TestStep.ConfigDirectory`, respectively. So I believe removing `.json` files will be ok from this perspective.

Do you know any way in which a non-Terraform JSON file might end-up in the working directory, and if so, whether removing it would be likely to cause issues?

Comment (bflad) [reply internal/plugintest/working_dir.go:103]: Hmm, I think you're right to a certain degree. Maybe I was thinking of the opposite problem then -- let's say we're using `ConfigDirectory` across multiple steps and I have:

```
# step 1 directory
|- file.whateverextension
 \ main.tf

# step 2 directory
\ main.tf
```

We should be preserving things necessary between steps in the working directory such as the `terraform.tfstate`, but removing the rest of the working directory files/directories. The configuration copying would overwrite any lingering files between steps, so in that regard its okay, but I think it would issue if the configuration was using something like `fileset()` instead of referencing thing with `file()` directory _and_ expecting those lingering files to be removed between steps. Let's treat this as something to follow up with rather than holding this up further though!

**Review by bflad — APPROVED** [review]

File: internal/plugintest/working_dir.go
Comment (bendbennett) [reply internal/plugintest/working_dir.go:103]: Thanks for the additional context. Have created a follow-up issue - [Consider removal of all files copied from ConfigDirectory between sequential test steps](https://github.com/hashicorp/terraform-plugin-testing/issues/172#top)

**Pull request merged**
